bayes_kit/ess.py

```diff
@@ -1,66 +1,26 @@
 import numpy as np
 import numpy.typing as npt
+import bayes_kit.autocorr as autocorr
 
 FloatType = np.float64
 IntType = np.int64
 VectorType = npt.NDArray[FloatType]
 
-def autocorr_fft(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function calls a fast Fourier transform (FFT).
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    size = 2 ** np.ceil(np.log2(2 * len(chain) - 1)).astype("int")
-    var = np.var(chain)
-    ndata = chain - np.mean(chain)
-    fft = np.fft.fft(ndata, size)
-    pwr = np.abs(fft) ** 2
-    N = len(ndata)
-    acorr = np.fft.ifft(pwr).real / var / N
-    return acorr
-
-def autocorr_np(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function delegates to the Numpy `correlation()` function.
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    chain_ctr = chain - np.mean(chain)
-    N = len(chain_ctr)
-    acorrN = np.correlate(chain_ctr, chain_ctr, "full")[N - 1 :]
-    return acorrN / N
-
-def autocorr(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function delegates to `autocorr_fft`.
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    return autocorr_fft(chain)
-
-def first_neg_pair_start(chain: VectorType) -> IntType:
-    """
-    Return the index of first element of the sequence whose sum with the following
-    element is negative, or the length of the sequence if there is no such element.
+def _end_pos_pairs(chain: VectorType) -> IntType:
+    """Return the index of first element of the sequence whose sum with
+
+    the following element is negative, or the length of the sequence if
+    there is no such element.  Equivalently, this is the upper bound for
+    the range of indexes that come in positive pairs starting from 0.
+    For exmaple, `_end_pos_pairs([1, -.5, .25, -.3]) == 2` because the
+    first pair (1, -.5) is positive whereas the second pair (.25 + -.3)
+    is negative.  Contrast this with `_end_pos_pairs([1, -.5, .25, -.1])
+    = 4`, in the case where the second pair (.25, -.1) is positive.
+    Trailing odd elements are ignored, so `_end_pos_pairs([1, -.5, .25, -.3, .05]) == 2`
+    and `_end_pos_pairs([1, -.5, .25, -.1, .05]) = 4`.
     
     Parameters:
-    chain: input sequence
+    chain: input sequence (typically of autocorrelations)
 
     Return:
     index of first element whose sum with following element is negative, or
```

Comment:
As above, I think it'd be best to call out that these are even-indexed elements only (and the function can never return an odd number).

Reply:
fixed along with last one